marine.c

```diff
@@ -966,6 +977,374 @@ WS_DLL_PUBLIC void marine_free(marine_result *ptr) {
+            g_array_append_val(new_children, child);
+        } else if (indices->len > 1) {
+            join_duplicate_fields(field, new_children, indices, key);
+        }
+        g_array_free(indices, TRUE);
+    }
+
+    g_array_free(field->children, TRUE);
+    field->children = new_children;
+    g_free(keys);
+    g_hash_table_destroy(indices_by_name);
+}
+
+static void
+marine_packet_field_load_children(marine_packet_field *tree, proto_node *node) {
+    struct marine_extract_proto_tree_data child_proto_data = {
+            .field = tree,
+            .child_name_set = g_hash_table_new(g_str_hash, g_str_equal),
+            .has_duplicate_children = FALSE
+    };
+    proto_tree_children_foreach(node, proto_tree_load_onto_packet, &child_proto_data);
+    if (child_proto_data.has_duplicate_children) {
+        marine_packet_field_join_children_with_duplicate_names(tree);
+    }
+    g_hash_table_destroy(child_proto_data.child_name_set);
+}
+
+static void marine_setup_wtap_rec(wtap_rec *rec, int len, int wtap_encap) {
+    // Fake the rec structure for internal dissection
+    rec->rec_type = REC_TYPE_PACKET;
+    rec->presence_flags = WTAP_HAS_CAP_LEN;
+    rec->rec_header.packet_header.caplen = len;
+    rec->rec_header.packet_header.len = len;
+    rec->rec_header.packet_header.pkt_encap = wtap_encap;
+    rec->rec_header.ft_specific_header.record_len = len;
+    rec->rec_header.ft_specific_header.record_type = len;
+    rec->rec_header.syscall_header.record_type = len;
+    rec->rec_header.syscall_header.byte_order = len;
+}
+
+static epan_dissect_t *
+marine_epan_dissect_new(capture_file *cf) {
+    /* The protocol tree will be "visible", i.e., printed, only if we're
+       printing packet details, which is true if we're printing stuff
+       ("print_packet_info" is true) and we're in verbose mode
+       ("packet_details" is true). */
+    epan_dissect_t *edt = epan_dissect_new(cf->epan, TRUE, TRUE);
+    reset_epan_mem(cf, edt, 1, 1);
+    return edt;
+}
+
+static void
+marine_epan_dissect_run(capture_file *cf, epan_dissect_t *edt, wtap_rec *rec, Buffer *buf, int len) {
+    frame_data fdata;
+    column_info *cinfo;
+
+    cf->count++;
+    marine_frame_data_init(&fdata, cf->count, len);
+
+    prime_epan_dissect_with_postdissector_wanted_hfids(edt);
+
+    col_custom_prime_edt(edt, &cf->cinfo);
+    cinfo = NULL;
+
+    if (cf->provider.ref == &fdata) {
+        ref_frame = fdata;
+        cf->provider.ref = &ref_frame;
+    }
+
+    epan_dissect_run_with_taps(edt,
+                               cf->cd_t,
+                               rec,
+                               frame_tvbuff_new_buffer(&cf->provider, &fdata, buf),
+                               &fdata,
+                               cinfo);
+
+    frame_data_set_after_dissect(&fdata, &cum_bytes);
+    prev_dis_frame = fdata;
+    cf->provider.prev_dis = &prev_dis_frame;
+    frame_data_destroy(&fdata);
+}
+
```

Comment:
Remove extra line (also in a few more functions)

Reply:
Removed.